Add article on using Dockerfiles/Compose, formatting tweaks and clarifications (#107)
devcontainers/devcontainers.github.io: _posts/2022-12-16-dockerfiles.md

---
layout: post
title:  "Using Images, Dockerfiles, and Docker Compose"
author: "@chuxel"
authorUrl: https://github.com/chuxel
---

> When creating a development container, you have a variety of different ways to customize your environment like ["Features"](/features) or [lifecycle scripts](implementors/json_reference/#lifecycle-scripts). However, if you a familiar with containers, you may want to use a [Dockerfile](#dockerfile) or [Docker Compose / Compose](#docker-compose) to customize your environment. This article will walk through how to use these formats with the Dev Container spec.

Reviewer: ```suggestion
When creating a development container, you have a variety of different ways to customize your environment like ["Features"](/features) or [lifecycle scripts](implementors/json_reference/#lifecycle-scripts). However, if you are familiar with containers, you may want to use a [Dockerfile](#dockerfile) or [Docker Compose / Compose](#docker-compose) to customize your environment. This article will walk through how to use these formats with the Dev Container spec.
```

## <a href="dockerfile" name="dockerfile" class="anchor">  Using a Dockerfile </a>

To keep things simple, many [Dev Container Templates](/templates) use container image references.

```json
{

>     "image": "mcr.microsoft.com/vscode/devcontainers/base:ubuntu"

Reviewer 2: ```suggestion
    "image": "mcr.microsoft.com/devcontainers/base:ubuntu"
```

}
```

> However, [Dockerfiles](https://docs.docker.com/engine/reference/builder/) are a great way to extend images and are a great way to add additional native OS packages or making minor edits to the OS image. You can reuse any Dockerfile, but let's walk through how to create one from scratch.

Reviewer: ```suggestion
However, [Dockerfiles](https://docs.docker.com/engine/reference/builder/) are a great way to extend images, add additional native OS packages, or make minor edits to the OS image. You can reuse any Dockerfile, but let's walk through how to create one from scratch.
```

First, add a file named `Dockerfile` next to your `devcontainer.json`. For example:

```Dockerfile

> FROM mcr.microsoft.com/vscode/devcontainers/base:ubuntu

Reviewer 2: ```suggestion
FROM mcr.microsoft.com/devcontainers/base:ubuntu
```

# Install the xz-utils package
RUN apt-get update && apt-get install -y xz-utils
```

Next, remove the `image` property from `devcontainer.json` (if it exists) and add the `build` and `dockerfile` properties instead:

```json
{
    "build": {

>         // Path is relataive to the devcontainer.json file.

Reviewer 2: Should we also add `context` property?

eg.

```
	"build": {
		// Sets the run context to one level up instead of the .devcontainer folder.
		"context": "..",
		// Path is relataive to the devcontainer.json file.
		"dockerfile": "Dockerfile"
	}
```

Author: For this article I wanted to stay with the bare minimum, and then later it references the properties section.

        "dockerfile": "Dockerfile"
    }
}
```

That's it! When you start up your Dev Container, the Dockerfile will be automatically built with no additional work. See [Dockerfile scenario reference](implementors/json_reference/#image-specific) for more information on other related devcontainer.json properties.

> ### Iterating on an image that includes Dev Container metadata

Reviewer: Can we make this a clickable heading link?

Better yet, you can can use a Dockerfile as a part of authoring an image you can share with others. You can even **add Dev Container settings and metadata right into the image itself**. This avoids having to duplicate config and settings in multiple devcontainer.json files and keeps them in sync with your images! 

See the reference on **[pre-building](/implementors/reference/#prebuilding)** to learn more!

## <a href="docker-compose" name="docker-compose" class="anchor">  Using a Dockerfile </a>

[Docker Compose](https://docs.docker.com/compose/) is a great way to define a multi-container development environment. Rather than adding things like databases or redis to your Dockerfile, you can reference existing images for these services and focus your Dev Container's content on tools and utilities you need for development.

> ### Using an image with Docker Compose

Reviewer: Can we make this a link too?

As mentioned in the Dockerfile section, to keep things simple, many [Dev Container Templates](/templates) use container image references.

```json
{

>     "image": "mcr.microsoft.com/vscode/devcontainers/base:ubuntu"

Reviewer 2: ```suggestion
    "image": "mcr.microsoft.com/devcontainers/base:ubuntu"
```

I bet you are so used to using the old image paths. hehe 😅

Author: Yep - total force of habit. Not even thinking when I type. lol

}
```

Let's create a `docker-compose.yml` file next to your `devcontainer.json` that references the same image and includes a PostgreSQL database:

```yaml
version: '3.8'
services:
  devcontainer:

>     image: mcr.microsoft.com/vscode/devcontainers/base:ubuntu

Reviewer 2: ```suggestion
    image: mcr.microsoft.com/devcontainers/base:ubuntu
```

    restart: unless-stopped
    volumes:
      - postgres-data:/var/lib/postgresql/data
    environment:
      POSTGRES_PASSWORD: postgres
      POSTGRES_USER: postgres
      POSTGRES_DB: postgres

volumes:
  postgres-data:
```

In this example:
-  `../..:/workspaces:cached` mounts the workspace folder from the local source tree into the Dev Container.
- `network_mode: service:db` puts the Dev Container on the same network as the database, so that it can access it on `localhost`.
- The `db` section uses the [Postgres](https://hub.docker.com/_/postgres) image with a few settings.

Next, let's configure devcontainer.json to use it.

```json
{
    "dockerComposeFile": "docker-compose.yml",
    "service": "devcontainer",
    "workspaceFolder": "/workspaces/${localWorkspaceFolderBasename}"
}
```

In this example:
- `service` indicates which service in the `docker-compose.yml` file is the Dev Container.
- `dockerComposeFile` indicates where to find the `docker-compose.yml` file.
- `workspaceFolder` indicates where to mount the workspace folder. This corresponds to a sub-folder under the mount point from `../..:/workspaces:cached` in the `docker-compose.yml` file.

That's it!

> ### Using a Dockerfile with Docker Compose

Reviewer: Make a link?

You can also combine these scenarios and use Dockerfile with Docker Compose. This time we'll update `docker-compose.yml` to reference the Dockerfile by replacing `image` with a similar `build` section:

```yaml
version: '3.8'
services:
  devcontainer:
    build: 
      context: .
      dockerfile: Dockerfile
    volumes:
      - ../..:/workspaces:cached      
    network_mode: service:db

  db:
    image: postgres:latest
    restart: unless-stopped
    volumes:
      - postgres-data:/var/lib/postgresql/data
    environment:
      POSTGRES_PASSWORD: postgres
      POSTGRES_USER: postgres
      POSTGRES_DB: postgres

volumes:
  postgres-data:
```

> Finally, as in the Dockerfile example, you can use this same setup to author a Dev Container image you can share with other and add Dev Container settings and metadata right into the image itself. 

Reviewer: ```suggestion
Finally, as in the Dockerfile example, you can use this same setup to author a Dev Container image you can share with others and add Dev Container settings and metadata right into the image itself. 
```
